html/captive-portal/lib/captiveportal/PacketFence/Controller/TLSProfile.pm

```diff
@@ -0,0 +1,180 @@
+
+=head1 NAME
+
+captiveportal::PacketFence::Controller::WirelessProfile - Catalyst Controller
+
+=head1 DESCRIPTION
+
+Catalyst Controller.
+
+=head1 METHODS
+
+=cut
+
+=head2 index
+
+=cut
+
+sub index : Path : Args(0) {
+    my ( $self, $c ) = @_;
+    my $username = $c->session->{username};
+    my $logger  = get_logger;
+    my $profile = $c->profile;
+    my $request = $c->request;
+    my $mac = $c->portalSession->clientMac;
+    my $provisioner = $c->profile->findProvisioner($mac);
+    $provisioner->authorize($mac) if (defined($provisioner));
+    $c->stash(
+        post_uri            => '/tlsprofile/cert',
+        destination_url     => 'eap-profile.html',
+        certificate_cn      => $request->param_encoded("certificate_cn"),
+        certificate_pwd     => $request->param_encoded("certificate_pwd"),
+        certificate_email   => lc( $request->param_encoded("certificate_email")),
+        service             => $request->param_encoded("service"),
+        profile_list        => $Config{'pki'}{'profile'},
+        template            => 'pki.html',
+        provisioner         => $provisioner,
+        username            => $username,
+        );
+}
+sub cert_p12 : Path('/eap-profile.html') : Args(0) {
+    my ( $self, $c ) = @_;
+    my $cert_data = $c->response->body($c->stash->{'cert_content'});
+    #my $headers = $c->response->headers;
+    #$headers->content_type('application/x-pkcs12');
+    my $certname = $c->stash->{'certificate_cn'} . "p12";
+    #$headers->header( 'Content-Disposition', "attachment; filename=\"$certname\"" );
+    open FH, "> $cert_dir/$certname" or die $!;
+    print FH "$cert_data\n";
+}
+
+
+sub get_cert : Private {
+    use bytes;
+    my ($self,$c,) = @_;
+    my $stash = $c->stash;
+    my $uri = $Config{'pki'}{'uri'};
+    my $username = $Config{'pki'}{'username'};
+    my $password = $Config{'pki'}{'password'};
+    my $email = $stash->{'certificate_email'};
+    my $dot1x_username = $stash->{'certificate_cn'};
+    my $organisation = $Config{'pki'}{'organisation'};
+    my $state = $Config{'pki'}{'state'};
+    my $profile = $Config{'pki'}{'profile'};
+    my $country = $Config{'pki'}{'country'};
+    my $certpwd = $stash->{'certificate_pwd'};
+    my $response = '';
+    my $curl = WWW::Curl::Easy->new; #$self->curl($function);
+    my $request = "username=$username&password=$password&cn=$dot1x_username&mail=$email&organisation=$organisation&st=$state&country=$country&profile=$profile&pwd=$certpwd";
+    my $response_body = '';
+    $curl->setopt(CURLOPT_POSTFIELDSIZE,length($request));
+    $curl->setopt(CURLOPT_POSTFIELDS, $request);
+    $curl->setopt(CURLOPT_WRITEDATA, \$response_body);
+    $curl->setopt(CURLOPT_HEADER, 0);
+    $curl->setopt(CURLOPT_DNS_USE_GLOBAL_CACHE, 0);
+    $curl->setopt(CURLOPT_NOSIGNAL, 1);
+    $curl->setopt(CURLOPT_URL, $uri);
+  
+    # Starts the actual request
+    my $curl_return_code = $curl->perform;
```

Comment:
It's nice to store the return code, but you should use it to check whether this request was a success or a failure. 

If the PKI is down, you're not handling it and you'll be rendering garbage to the user.